feat: add/remove fields from kanban board

hrwX opened the pull request:
- Current Kanban Implementation didn't allow users to customize the fields visible in the card.
- This PR adds the ability to add/remove fields from the kanban via Kanban Settings and overrides the List Settings for Kanban View.


https://user-images.githubusercontent.com/14891507/162844294-dcd7bbe2-5483-4e7f-ab75-2cbd32b48c29.mov

https://discuss.erpnext.com/t/improved-kanban-board/88414

> no-docs

hrwX committed:
feat: add fields to kanban

hrwX committed:
fix: translate labels

hrwX committed:
Merge branch 'develop' into kanban_fields

hrwX committed:
chore: sider fixes

hrwX committed:
Merge branch 'kanban_fields' of https://github.com/hrwX/frappe into kanban_fields

barredterra commented on frappe/public/js/frappe/views/kanban/kanban_board.js:
```diff
@@ -640,12 +640,25 @@ frappe.provide("frappe.views");
+
+			JSON.parse(cur_list.board.fields || "[]").forEach(field => {
+				if (card.doc[field.fieldname] && card.doc[field.fieldname] !== card.title) {
+					fields.push(`<span>${__(field.label)}: ${__(card.doc[field.fieldname])}</span>`);
```
@hrwX can we show the formatted value here?

barredterra replied:
Also, why construct the HTML here? I'd suggest to move the HTML structure into `kanban_card.html`.

barredterra commented on frappe/public/js/frappe/views/kanban/kanban_board.js:
```diff
@@ -640,12 +640,25 @@ frappe.provide("frappe.views");
+
+			JSON.parse(cur_list.board.fields || "[]").forEach(field => {
+				if (card.doc[field.fieldname] && card.doc[field.fieldname] !== card.title) {
+					fields.push(`<span>${__(field.label)}: ${__(card.doc[field.fieldname])}</span>`);
```
@hrwX can we show the formatted value here?

barredterra reviewed (commented):
@hrwX some thoughts, let me know your take:

- In List Settings I can only select fields that have "In List View" enabled. Maybe this would be useful for Kanban Settings as well? This way developers and system managers can choose useful defaults in advance. Also the information required in Kanban Card and List View will probably be very similar. (requested by the customer)
- I would personally find the following useful as well:
    - Add standard fields to selection: `name`, `owner`, `creation`, `modified`, `modified_by`. This is valuable information, but hard to access deliberately at the moment.
        - In this case the creation date that used to be shown in the cards should become configurable. Remove it from the hardcoded values and add a patch to set it as a configured field for existing boards.
    - Hide internal fields from treeview: `lft`, `rgt`, `old_parent`. Probably completely useless to everybody.
- Store configured fields in a proper child table instead of JSON dumped in a text field
- Show formatted value instead of raw data

barredterra commented on frappe/public/js/frappe/views/kanban/kanban_view.js:
```diff
@@ -26,6 +28,12 @@ frappe.views.KanbanView = class KanbanView extends frappe.views.ListView {
 	}
 
+	get_fields_in_list_view() {
+		let fields = JSON.parse(this.board.fields).map(e => e.fieldname);
```
```suggestion
		let fields = JSON.parse(this.board.fields || "[]").map(e => e.fieldname);
```

I got an error for an existing board where `this.board.fields` was empty.

barredterra commented on frappe/public/js/frappe/views/kanban/kanban_card.html:
```diff
@@ -11,6 +11,10 @@
+					<div class="kanban-card-doc text-muted">
+						{{ doc_content }}
+					</div>
 					<div class="kanban-card-creation">
 						{{ creation }}
 					</div>
```
```suggestion
```

barredterra commented on frappe/public/js/frappe/views/kanban/kanban_card.html:
```diff
@@ -11,6 +11,10 @@
 					</div>
+					<br>
+					<div class="kanban-card-doc text-muted">
+						{{ doc_content }}
```
Something like (untested)

```suggestion
						{% for field in fields %}
						    <span>{{ field.fieldname }}: {{ frappe.get_formatted(field.value, field.fieldtype) }}</span><br>
						{% endfor %}
```

hrwX committed:
Merge branch 'develop' of https://github.com/frappe/frappe into kanban_fields

hrwX committed:
fix: stringified list fallback in case fields is empty

Co-authored-by: Raffael Meyer <14891507+barredterra@users.noreply.github.com>

hrwX committed:
Merge branch 'kanban_fields' of https://github.com/hrwX/frappe into kanban_fields

hrwX commented:
> * ay developers and system managers can choose useful defaults in adv



> @hrwX some thoughts, let me know your take:
> 
> * In List Settings I can only select fields that have "In List View" enabled. Maybe this would be useful for Kanban Settings as well? This way developers and system managers can choose useful defaults in advance. Also the information required in Kanban Card and List View will probably be very similar. (requested by the customer)
> * I would personally find the following useful as well:
>   
>   * Add standard fields to selection: `name`, `owner`, `creation`, `modified`, `modified_by`. This is valuable information, but hard to access deliberately at the moment.
>     
>     * In this case the creation date that used to be shown in the cards should become configurable. Remove it from the hardcoded values and add a patch to set it as a configured field for existing boards.
>   * Hide internal fields from treeview: `lft`, `rgt`, `old_parent`. Probably completely useless to everybody.
> * Store configured fields in a proper child table instead of JSON dumped in a text field
> * Show formatted value instead of raw data

- In List Settings, by default will show all the fields in listview via meta, we can add the listview fields as defaults.
- Adding the standard fields to the selector
- Kanban Board doctype shouldn't be accessible to the user in the first place via global search, will change it.

hrwX committed:
fix: add std fields

shariquerik commented:
@hrwX the sequence is resetting after adding a new field. Check video

https://user-images.githubusercontent.com/30859809/159234576-63b6f63c-af44-4304-8f20-a4262f4d1d03.mov


Also, sometimes while adding a new field it is not displayed after saving we have to reload the page.

hrwX committed:
Merge branch 'develop' into kanban_fields

hrwX committed:
Merge branch 'develop' of https://github.com/frappe/frappe into kanban_fields

hrwX committed:
fix: do not reset fields order

hrwX committed:
Merge branch 'kanban_fields' of https://github.com/hrwX/frappe into kanban_fields

hrwX commented:
> @hrwX the sequence is resetting after adding a new field. Check video
> 
>  KanbanSettingBug.mov 
> Also, sometimes while adding a new field it is not displayed after saving we have to reload the page.

@shariquerik This has been fixed. Also can you replicate and record the bug where field not being visible without refresh ?

hrwX committed:
Merge branch 'develop' into kanban_fields

shariquerik commented:
@hrwX sequence issue is resolved but I found one more issue. If we try to add multiple fields it only adds one check video.

https://user-images.githubusercontent.com/30859809/161901580-2822c5d6-57bf-483b-acd5-01f02a9d1e0b.mov

>@shariquerik This has been fixed. Also can you replicate and record the bug where field not being visible without refresh ?

https://user-images.githubusercontent.com/30859809/161901947-cebc8d58-da6a-4135-b6e8-6b7af0b949b2.mov

barredterra committed:
style: format with prettier

barredterra committed:
refactor: me -> this, style, property -> variable/constant

barredterra committed:
fix: select multiple fields, update board

barredterra commented:
@shariquerik thanks for testing. The issues you found should be resolved:

https://user-images.githubusercontent.com/14891507/162334211-82a167d4-6237-4aa1-809c-2c6e467c3ec6.mov

barredterra committed:
feat: save only fieldnames, show formatted value

Refactor:
- Replace `me` with `this`
- Remove useless parameters

barredterra committed:
feat: hide Kanban Board in search

barredterra committed:
Merge branch 'develop' into kanban_fields

barredterra committed:
fix: ignore treeview fields

barredterra committed:
fix: update fields

barredterra committed:
feat: remove unfunctional reset button

barredterra committed:
fix: card layout

- card title links to form
- whole card can be grabbed
- increase height of empty columns to serve as target for drag-and-drop

barredterra committed:
feat: ignore unsupported fieldtypes

barredterra committed:
perf: less json parsing

barredterra committed:
style: format

barredterra committed:
refactor: remove unused method

shariquerik commented:
@barredterra The default field is missing in Kanban Setting
<img width="993" alt="image" src="https://user-images.githubusercontent.com/30859809/162678523-576621dc-c22a-4b52-9902-5693806414e0.png">
It should have **Subject** field
Before: 
<img width="624" alt="image" src="https://user-images.githubusercontent.com/30859809/162678747-04be0f74-a716-4cb3-b73d-ca44dd881da9.png">

I think because of this there is one more issue. If you remove all fields from the field selection list (uncheck all instead of deleting it from selected fields list) it does not allow to save

barredterra commented:
@shariquerik I removed the default field on purpose because it doesn't add functionality, just makes the code more complicated.

I'll have a look at unchecking all fields

barredterra committed:
fix: error when unchecking all fields

barredterra commented:
> If you remove all fields from the field selection list (uncheck all instead of deleting it from selected fields list) it does not allow to save

@shariquerik this should be resolved now.

barredterra committed:
Merge branch 'develop' into kanban_fields

barredterra committed:
Merge branch 'develop' into kanban_fields

barredterra committed:
style: run pre-commit

barredterra committed:
fix: scss

- refactor structure
- add display: inline for doc-content to prevent linebreak

barredterra committed:
feat: allow only fields that are in list view

We don't load data for other fields anyways, so they would be empty.

barredterra commented:
KanbanView extends the ListView, which loads only data for fields that are marked as "In List View". Even if we allowed to select other fields for the KanbanView, they would hold no data. Two possible solutions:

(a) modify KanbanView to load data for all fields that have been selected in KanbanSettings,
(b) modify KanbanSettings to only allow selecting of fields that are loaded by ListView

I went for (b) as that was the suggested design in https://github.com/frappe/frappe/pull/16257#pullrequestreview-909255285 anyways.

gavindsouza commented:
Some changes I wanted to make in https://github.com/frappe/frappe/pull/16509 but decided against since that would increase conflicts with this PR:

* Make Kanban Board's filters visible in the Document Form (They're hidden at the moment)
* Track Changes for Kanban Board

Wanted to make these changes since users are allowed to save filters which change the view for everyone else. There should be a record of the changes.

@hrwX Can you include these changes in this PR?

barredterra committed:
Merge branch 'develop' into kanban_fields

barredterra committed:
feat(Kanban Board): unhide filters, track changes

barredterra committed:
feat(Kanban Board): permissions

- System Manager: RWCD
- All: R, WCD if owner

barredterra commented:
@gavindsouza, done. Also restricted permissions so that users can only modify their own Boards.

barredterra committed:
feat: allow to toggle labels

barredterra committed:
refactor: function name

barredterra committed:
fix: sider

barredterra committed:
fix: dump json when necessary

barredterra committed:
Merge branch 'develop' into kanban_fields

shariquerik commented:
@barredterra Can we add a UI test for this

shariquerik committed:
Merge branch 'develop' into kanban_fields

hrwX committed:
Merge branch 'develop' into kanban_fields

shariquerik commented:
@barredterra While I was writing a UI test, I found one small issue. Every time I edit the fields for kanban, I have to select the show labels checkbox. It should also be saved in user settings.

I will have to wait for these changes to add the UI test.

barredterra commented:
@shariquerik yes, I noticed that as well, but couldn't find the cause yet. We use the standard `set_values` method, that is used everywhere in the framework, to sync the Kanban Settings (`this.settings`) with the dialog's field group:

https://github.com/frappe/frappe/blob/d026647c38ffc40272cb3c86483f077d86ab0838/frappe/public/js/frappe/views/kanban/kanban_settings.js#L44

In this case it reliably fails for the checkbox value. However, I couldn't reproduce it for other dialogs.

barredterra commented:
@shariquerik I found the change that introduced this problem: https://github.com/frappe/frappe/issues/16643

Let's continue with this PR and solve the issue separately.

shariquerik committed:
test: Cypress test for kanban board

shariquerik committed:
Merge branch 'develop' into kanban_fields

shariquerik committed:
test: Updated test name and fixed failing test

ankush commented:
Wohoo! 🎉  this looks great @hrwX @barredterra @shariquerik

shariquerik commented:
@Mergifyio backport version-13-hotfix
